agent/consul/authmethod/authmethods.go

```diff
@@ -42,16 +46,33 @@ type Validator interface {
 	// Returns auth method specific metadata suitable for the Role Binding
 	// process as well as the desired enterprise meta for the token to be
 	// created.
-	ValidateLogin(loginToken string) (map[string]string, *structs.EnterpriseMeta, error)
+	ValidateLogin(ctx context.Context, loginToken string) (*Identity, error)
+
+	// Stop should be called to cease any background activity and free up
+	// resources.
+	Stop()
+}
+
+type Identity struct {
+	// SelectableFields is the format of this Identity suitable for selection
+	// with a binding rule.
+	SelectableFields interface{}
 
-	// AvailableFields returns a slice of all fields that are returned as a
-	// result of ValidateLogin. These are valid fields for use in any
-	// BindingRule tied to this auth method.
-	AvailableFields() []string
+	// ProjectedVars is the format of this Identity suitable for interpolation
+	// in a bind name within a binding rule.
+	ProjectedVars map[string]string
 
-	// MakeFieldMapSelectable converts a field map as returned by ValidateLogin
-	// into a structure suitable for selection with a binding rule.
-	MakeFieldMapSelectable(fieldMap map[string]string) interface{}
+	*structs.EnterpriseMeta
+}
+
+// ProjectedVarNames returns just the keyspace of the ProjectedVars map.
+func (i *Identity) ProjectedVarNames() []string {
+	v := make([]string, 0, len(i.ProjectedVars))
+	for k, _ := range i.ProjectedVars {
+		v = append(v, k)
+	}
+	sort.Strings(v)
```

Comment:
Why must we sort? I see it done in the testing as well?

Reply:
Technically it's not needed for anything other than testing. I'll move the sorting to just happen in that test assertion helper.